File: src/services/ScheduleService.ts
import { Prescription, PrescriptionSchedule } from '../types';
import { ScheduleModel } from '../models/ScheduleModel';

export class ScheduleService {
  private scheduleModel: ScheduleModel;

  constructor() {
    this.scheduleModel = new ScheduleModel();
  }

  async generateSchedule(prescription: Prescription): Promise<PrescriptionSchedule[]> {
    const schedules: PrescriptionSchedule[] = [];
    const startDate = new Date(prescription.startDate);
    const endDate = new Date(prescription.endDate);
    
    for (let date = new Date(startDate); date <= endDate; date.setDate(date.getDate() + 1)) {
      const currentDate = date.toISOString().split('T')[0];
      
      for (let hour = 0; hour < 24; hour += prescription.frequencyHours) {
        const [startHour, startMinute] = prescription.startTime.split(':').map(Number);
        const scheduledHour = (startHour + hour) % 24;
        
        if (hour + startHour >= 24 && date.getTime() === endDate.getTime()) {
          break;
        }
        
        const scheduledTime = `${scheduledHour.toString().padStart(2, '0')}:${startMinute.toString().padStart(2, '0')}`;
        
        const schedule: PrescriptionSchedule = {
          prescriptionId: prescription.prescriptionId,
          patientId: prescription.patientId,
          scheduledDate: currentDate,
          scheduledTime,
          dosage: prescription.dosage,
          unit: prescription.unit,
          status: 'PENDING'
        };
        
        schedules.push(schedule);
      }
    }
    
    for (const schedule of schedules) {
      await this.scheduleModel.createScheduleEntry(schedule);
    }
    
    return schedules;
  }

  async getPrescriptionSchedule(prescriptionId: string): Promise<PrescriptionSchedule[]> {
    return this.scheduleModel.getScheduleByPrescription(prescriptionId);
  }

  async getPatientSchedule(patientId: string, date?: string): Promise<PrescriptionSchedule[]> {
    return this.scheduleModel.getPatientSchedule(patientId, date);
  }

  async updateScheduleStatus(
    prescriptionId: string, 
    scheduledDate: string, 
    scheduledTime: string, 
    status: 'PENDING' | 'TAKEN' | 'MISSED'
  ): Promise<PrescriptionSchedule | null> {
    return this.scheduleModel.updateScheduleStatus(prescriptionId, scheduledDate, scheduledTime, status);
  }

  async deleteScheduleForPrescription(prescriptionId: string): Promise<void> {
    return this.scheduleModel.deleteScheduleByPrescription(prescriptionId);
  }
}

File: kotlin-prescription-service/src/main/kotlin/com/prescription/service/ScheduleService.kt
package com.prescription.service

import com.prescription.domain.Prescription
import com.prescription.domain.PrescriptionSchedule
import com.prescription.domain.ScheduleStatus
import com.prescription.repository.ScheduleRepository
import jakarta.inject.Singleton
import java.time.LocalDate
import java.time.LocalTime

@Singleton
open class ScheduleService(
    private val scheduleRepository: ScheduleRepository
) {

    fun generateSchedule(prescription: Prescription): List<PrescriptionSchedule> {
        val schedules = mutableListOf<PrescriptionSchedule>()
        var currentDate = prescription.startDate
        
        while (!currentDate.isAfter(prescription.endDate)) {
            var currentHour = 0
            
            while (currentHour < 24) {
                val scheduleTime = prescription.startTime.plusHours(currentHour.toLong())
                
                // Skip if this would be the next day
                if (scheduleTime.isBefore(prescription.startTime) && currentHour > 0) {
                    break
                }
                
                // Skip if this goes beyond the end date
                if (currentDate == prescription.endDate && currentHour + prescription.startTime.hour >= 24) {
                    break
                }
                
                val adjustedTime = if (scheduleTime.hour >= 24) {
                    LocalTime.of(scheduleTime.hour % 24, scheduleTime.minute)
                } else {
                    scheduleTime
                }
                
                val schedule = PrescriptionSchedule(
                    prescriptionId = prescription.prescriptionId,
                    patientId = prescription.patientId,
                    scheduledDate = currentDate,
                    scheduledTime = adjustedTime,
                    dosage = prescription.dosage,
                    unit = prescription.unit,
                    status = ScheduleStatus.PENDING
                )
                
                schedules.add(scheduleRepository.createScheduleEntry(schedule))
                currentHour += prescription.frequencyHours
            }
            
            currentDate = currentDate.plusDays(1)
        }
        
        return schedules
    }

    fun getPrescriptionSchedule(prescriptionId: String): List<PrescriptionSchedule> {
        return scheduleRepository.getScheduleByPrescription(prescriptionId)
    }

    fun getPatientSchedule(patientId: String, date: String? = null): List<PrescriptionSchedule> {
        return scheduleRepository.getPatientSchedule(patientId, date)
    }

    fun updateScheduleStatus(
        prescriptionId: String,
        scheduledDate: String,
        scheduledTime: String,
        status: ScheduleStatus
    ): PrescriptionSchedule? {
        return scheduleRepository.updateScheduleStatus(prescriptionId, scheduledDate, scheduledTime, status)
    }

    fun deleteScheduleForPrescription(prescriptionId: String) {
        scheduleRepository.deleteScheduleByPrescription(prescriptionId)
    }
}